### ISSUE TITLE
Cannot access frames after a given point in large DCD files 

### ISSUE THREAD

**Issue by rebeccawalters95** [MDAnalysis/mdanalysis#3075]
## Expected behavior ##

I have an MDAnalysis universe (using a pdb and dcd file) where the trajectory is very large (2.5 GB) and has a lot of frames (25,000). 

```universe = mda.Universe(PDB, DCD)```

I am trying to trim this trajectory by selecting and then keeping every 10 frames using this line of code:

```universe.atoms.write('short.xyz', frames=f)```  where ```f = list(range(0,25000,10))```

I expect this to return a new trajectory called short.xyz which contains every 10th frame from the original (large) trajectory.

## Actual behavior ##

What actually happens is that when I try and write to a new trajectory, I get the following error...

```---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-25-79cfe83634e9> in <module>
----> 1 u.atoms.write('short.xyz', frames=f)

~\anaconda3\envs\BXD\lib\site-packages\MDAnalysis\core\groups.py in write(self, filename, file_format, filenamefmt, frames, **kwargs)
   3227                     current_frame = trj.ts.frame
   3228                     try:
-> 3229                         for _ in trj_frames:
   3230                             w.write(self.atoms)
   3231                     finally:

~\anaconda3\envs\BXD\lib\site-packages\MDAnalysis\coordinates\base.py in __iter__(self)
   1085     def __iter__(self):
   1086         for frame in self.frames:
-> 1087             yield self.trajectory._read_frame_with_aux(frame)
   1088 
   1089     def __getitem__(self, frame):

~\anaconda3\envs\BXD\lib\site-packages\MDAnalysis\coordinates\base.py in _read_frame_with_aux(self, frame)
   1610     def _read_frame_with_aux(self, frame):
   1611         """Move to *frame*, updating ts with trajectory, transformations and auxiliary data."""
-> 1612         ts = self._read_frame(frame)  # pylint: disable=assignment-from-no-return
   1613         for aux in self.aux_list:
   1614             ts = self._auxs[aux].update_ts(ts)

~\anaconda3\envs\BXD\lib\site-packages\MDAnalysis\coordinates\DCD.py in _read_frame(self, i)
    185         """read frame i"""
    186         self._frame = i - 1
--> 187         self._file.seek(i)
    188         return self._read_next_timestep()
    189 

MDAnalysis/lib/formats/libdcd.pyx in MDAnalysis.lib.formats.libdcd.DCDFile.seek()

OverflowError: Python int too large to convert to C long
```

The working hypothesis for the error thus far is: at some point, MDAnalysis asks the system to move a cursor to a specific place in a file. The position is given a long int which maximum is 2,147,483,647. If the position I request is greater than this value, it would explain the error.

## Code to reproduce the behavior ##

``` python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import TPR, XTC

u = mda.Universe(TPR, XTC)

# creating a big trajectory using MDAnalysis test suite, to avoid having to download 2.5 GB file
ag = u.select_atoms("all")
with mda.Writer('test.dcd', ag.n_atoms) as w:
    for _ in range(10000):
        for ts in u.trajectory:
            w.write(ag)

# creating new universe with the large trajectory
big_u = mda.Universe(TPR, 'test.dcd')

for time_step in range(big_u.trajectory.n_frames):
    print(big_u.trajectory[time_step])
```
^ this reproduces ```OverflowError: Python int too large to convert to C long``` 

## Current version of MDAnalysis ##

MDAnalysis version: 1.0.0
Python Version: Python 3.8.5
OS: Windows 10

**Comment by IAlibay** [comment]
thanks for raising this issue @rebeccawalters95 

I'm testing this locally to see if I can duplicate on linux.

Quick question, do you happen to know if you're running 32 bit or 64 bit python?

Probably the easiest way to check this is to do:

```python
import sys
print(sys.maxsize)
```

This should either return the value of a long int (32 bit) or a long long int (64 bit)

**Comment by rebeccawalters95** [comment]
thanks for testing it!

The output of ```sys.maxsize``` is ```9223372036854775807```

**Comment by IAlibay** [comment]
So I had to truncate to ~ 21 GB (my poor m.2 drive can't handle that much disk space), which is 38783 frames.

I didn't manage to replicate the error with that, can you confirm that this is also the case for you or do you also get failures if you set `range(3878)` instead of `range(10000)`?

**Comment by rebeccawalters95** [comment]
the code to reproduce the error gets to ```< Timestep 3752 with unit cell dimensions [79.98359 79.98359 79.98358 60.      60.      90.     ] >``` before crashing

**Comment by IAlibay** [comment]
@rebeccawalters95 so `range(376)` would be around the size limit where it would fail? (just checking so I can cut down the file size and try it out on my windows laptop)

**Comment by IAlibay** [comment]
Ok, so tried this on both linux and windows w/ 1.0.0

On linux: everything passes as expected.
On windows: I get the failure `OverflowError` at 3752.

I'm going to tag this as Windows specific unless someone else can reproduce it on other OS.

As per @rebeccawalters95's comments, I also suspect for some reason Window's compiler is defaulting into a `long int` instead of a `long long int` somewhere in libdcd.pyx 😕

**Comment by richardjgowers** [comment]
@rebeccawalters95 a hack around this might be to do the slice manually (the ugliness is making `i` loop from 0 to 9 then wrap around, then write every time it hits 0)

```python
with mda.Writer('out.xyz') as w:
    i = 0
    for ts in u.trajectory:
        i %= 10
        if not i:
            w.write(u.atoms)
        i += 1
```

From skimreading the code I think this avoids having any too large ints

**Comment by jbarnoud** [comment]
I could also reproduce the behaviour on windows. I came to the same conclusion as @IAlibay regarding the type issue in libdcd.pyx. More precisely, I think the error occurs here: https://github.com/MDAnalysis/mdanalysis/blob/222336396417c5d0ded2b461448bfbe856b4b99d/package/MDAnalysis/lib/formats/libdcd.pyx#L427

I computed the value of `offset` following a failure in @rebeccawalters95 code. It seems to overflow a long int indeed.

```
offset = big_u.trajectory._file._header_size + big_u.trajectory._file._firstframesize + big_u.trajectory._file._framesize * time_step
print(offset)  # 2148234364
print(offset > 2_147_483_647)  # True
```

Offset is supposed to be of type `fio_size_t` that is defined as `LONGLONG` in `fastio.h`: https://github.com/MDAnalysis/mdanalysis/blob/222336396417c5d0ded2b461448bfbe856b4b99d/package/MDAnalysis/lib/formats/include/fastio.h#L39 and of type `off_t` on other OSes: https://github.com/MDAnalysis/mdanalysis/blob/222336396417c5d0ded2b461448bfbe856b4b99d/package/MDAnalysis/lib/formats/include/fastio.h#L326

I also confirm that the error only occurs when seeking to a frame. Iterating does not encounter the issue and seems to be a suitable workaround @rebeccawalters95's problem. The issue remains, though.

**Comment by rebeccawalters95** [comment]
Thank you all for looking at the problem. I can confirm @richardjgowers workaround did indeed work - thanks very much!

**Comment by richardjgowers** [comment]
fixed in #4048
